### ISSUE TITLE
Needs a really important improvement

### ISSUE THREAD

**Issue by aurooxygen** [transmission/transmission#6554]
### What is the issue?

You can’t have this app download endlessly. The number of times I’ve lost HOURS of productivity because Transmission Mac kept downloading till there is litterall a couple of mb free disk space is insane.

PLEASE have a buffer setting where I can define the min free disk space before all downloads are halted. It is extraordinary that it is 2024 and this still doesn’t exist. I’m sorry you’re a NFP and donation based, but this is a really essential feature.

Additionally, when I type in TORRENT in Spotlight on Mac, please do something so this app shows up. I can never remember the name (just like I can’t remember to check my free space before I start downloading). 

Thank you for being the only viable alternative left to the mess that is utorrent Web.

### Which application of Transmission?

macOS app

### Which version of Transmission?

4.0.4

**Comment by killemov** [comment]
I agree. But also search before you post. Duplicate of #1835

**Comment by killemov** [comment]
@aurooxygen Yes, fixing of some issues seems to move at a glacial pace while others get picked up immediately. This issue is as old as Transmission is, so at least it is something the current devs have not introduced themselves. In the mean time you could put your download folder outside your system partition so that does not get FUBAR. I did the same on my Linux machine.

**Comment by frandroid** [comment]
> when I type in TORRENT in Spotlight on Mac, please do something so this app shows up. 

You can rename the app locally, to "Transmission (torrent).app" or something. :)

**Comment by porglezomp** [comment]
> Additionally, when I type in TORRENT in Spotlight on Mac, please do something so this app shows up. I

For developers: you can add "torrent" to `MDItemKeywords` in the Info.plist to make it known to Spotlight that the app is associated with torrents. ([Apple docs](https://developer.apple.com/documentation/coreservices/kmditemkeywords?language=objc), [Stackoverflow answer](https://stackoverflow.com/questions/15966923/how-to-fill-keywords-field-for-macosx-app-coded-in-xcode))

**Comment by Coeur** [comment]
Can you help to tell us how Apple apps are behaving in similar situations?
- Say, you have 101 MB free, and you download a 100 MB file with Safari?
- Say, you have 101 MB free, and you download a 100 MB app with AppStore?

Either there is a known Apple strategy to adopt, or it's a general macOS issue that can fail with any app already?

**Comment by killemov** [comment]
> In the mean time you could put your download folder outside your system partition so that does not get FUBAR.

**Comment by frandroid** [comment]
@aurooxygen  You could also just use an external drive... I mean if you're such a prolific torrent user you probably don't have enough of your laptop's storage capacity anyway.  You can keep your In Progress folder on your laptop but have the final destination be on the external drive...

**Comment by Coeur** [comment]
As a Transmission developer, I agree with @aurooxygen that something should be done in Transmission to avoid a system failure.
I just want to know first if Apple has examples in one of their apps (Safari, AppStore, Xcode, ...) that would show an example of good practice. Otherwise, if Apple doesn't have a good solution to that, then we're free to design a buffer settings with an arbitrary default value.
So if one wants to help here: kindly crash-test your Mac on macOS Sonoma with Safari, Mail, Notes, AppStore, Xcode or other native apps until you run out of disk space, then share the results: system freeze or does the app prevents you from risky downloading with some app-specific popup?

**Comment by Coeur** [comment]
After maybe 30 minutes (because I have a lot of free space):
```
$ mkfile 999g A_FILE.img
mkfile: (A_FILE.img removed) Write Error: No space left on device
```
So at least, that one command has native protection: it will detect when running out of space and will cancel the whole write operation.

**Comment by Coeur** [comment]
OK, I found the two apis we can use:
1. [NSBundleResourceRequestLowDiskSpaceNotification](https://developer.apple.com/documentation/foundation/nsbundleresourcerequestlowdiskspacenotification?language=objc) - we can use it to detect automatically when running low on disk space
2. [NSFileSystemFreeSize](https://developer.apple.com/documentation/foundation/file_system/about_apple_file_system?language=objc) - we can use it to tell how much space is left ([although it requires to declare NSPrivacyAccessedAPICategoryDiskSpace in `PrivacyInfo.xcprivacy`](https://developer.apple.com/documentation/bundleresources/privacy_manifest_files/describing_use_of_required_reason_api?language=objc)):
```swift
if let attributes = try? FileManager.default.attributesOfFileSystem(forPath: "/") {
    let availableFreeSpace = attributes[.systemFreeSize] 
}
```

**Comment by Coeur** [comment]
> I agree. But also search before you post. Duplicate of #1835

So, in the absolute, yes. But that old issue doesn't specify which OS is the issue occurring on.
And according to [this StackOverflow answer](https://stackoverflow.com/questions/38602652/how-to-detect-insufficient-disk-space-when-using-stdio-file-operations-in-c-c):
>The standard library has no functions for this.

So the macOS problem would only be resolved with an Apple Foundation api ([NSBundleResourceRequestLowDiskSpaceNotification](https://developer.apple.com/documentation/foundation/nsbundleresourcerequestlowdiskspacenotification?language=objc)), so I'll keep the present #6554 to track the macOS-specific issue, and leave the older issue for asking C++ folks to improve the standard library.

**Comment by killemov** [comment]
> > I agree. But also search before you post. Duplicate of #1835
> 
> So, in the absolute, yes. But that old issue doesn't specify which OS is the issue occurring on.

In this case it doesn't matter because the problem happens on all(?) OSes. The kicker is that TM itself becomes useless once a partition is full. You can't delete or move content with TM, you need to make some space manually before you can use TM again. Which is a pain if you don't have filesystem access. Just the mentioned tiny low water mark would fix this. Also the option of pre-allocating with all 1s could make a difference on filesystems that support sparse files.

**Comment by GaryElshaw** [comment]
I'm on the outside of this being a 'Transmission problem'. I think it's great that @Coeur is tackling this, but this, to me, is not a "really important improvement". 

Being observant of disk space isn't confined to torrents. It's an every-day consideration most people are mindful of when doing an OS update, an Xcode update, or any other hefty data exchange. I am not throwing shade on @aurooxygen, but blaming the client for being unobservant, is something i find to be a distraction from far greater project focus.

**Comment by killemov** [comment]
@GaryElshaw I mostly agree but TM being stuck when the disk is full kinda sucks.

**Comment by binarykitchen** [comment]
Hi folks, I'm new here and have raised ticket [#6740](https://github.com/transmission/transmission/issues/6740) which @tearfur linked, thank you.

Ok, I'm trying to catch up and am no expert in that field, yet, I feel something doesn't seem right. Where to start?

- Running transmission on an encrypted server these days is quite the standard.
- If disk space runs out on a server running transmission, you have to SSH into your server in root mode.
- Then enter some very long pass phrase from a paper you forgot in your office, while all that is quite difficult to remember, remount it and so on, which can be quite stressful.
- All that stress just because there is no such check, a pre-allocation check in Tranmission it seems.
- I come from macOS. uTorrent has this option. They did it and worked for me, now I switched to Ubuntu and experience this funny problem.
- Yeah it's bloody hard work, no criticism. Just trying to share.
- If all depends on some Apple API like @Coeur pointed to ([NSBundleResourceRequestLowDiskSpaceNotification](https://developer.apple.com/documentation/foundation/nsbundleresourcerequestlowdiskspacenotification?language=objc)),(https://developer.apple.com/documentation/foundation/nsbundleresourcerequestlowdiskspacenotification?language=objc) I would say, this is not an excuse we cannot solve this on Linux.

**Comment by killemov** [comment]
As I stated in #1835 a low water mark system seems to be the easiest solution: Stop any torrent that wants to allocate space that would get the free space below a configurable low water mark.

**Comment by Coeur** [comment]
OK, then please use that other issue (#1835) to implement the low watermark system on torrent allocations, and once that is implemented, I'll implement the UI part in macOS Transmission settings with present issue (#6554).

**Comment by ckerr** [comment]
Closing as duplicate of #1835
